core/src/main/java/org/hibernate/search/jsr352/internal/steps/lucene/EntityReader.java

```diff
@@ -242,7 +242,8 @@ private ScrollableResults buildScrollUsingHQL(StatelessSession ss, String HQL) {
 	private ScrollableResults buildScrollUsingCriteria(StatelessSession ss,
 			PartitionBound unit, Object checkpointId, JobContextData jobData) {
 
-		String idName = MassIndexerUtil.getIdName( entityType, session );
+		String idName = Projections.id().toString();
```

Comment:
### 1. Testing

> Do you want me to create an entity type MyEntity.class which has a customized name of the identifier and test it through the batch indexing job?

Yes, that's what I meant. I think the batch won't work if the id property is named `customId`, and that's what this issue is about.

Keep in mind that you could also have a non-id property named "id", while the actual id property (annotated with `@Id`) would be named `customId`:

```
@Entity
@Indexed
public class MyEntity implements Serializable {

    @Id
    @DocumentId
    private int customId; // id column is named sth other than id

    @Basic
    @Column(name = "id")
    private String id;  // property named "id", but is not the id!

    ...
}
```

I wouldn't recommend this for real-world applications, but it's entirely possible, and it's something we should address.

### 2. Ordering with an aliased projection on the ID

> As for Order.property( "aliased_id" ), this method Order#property(String) doesn't exist. (I'm using Hibernate ORM 5.1.0.Final).

Right, the method was `Order.asc`, not `Order.property`.

> So basically, I don't see how to use Alias and I don't know how to change the code. Could you give me more detail, ideally some lines of code?

The purpose of an alias is to define a string that can later be used to refer to the projection that was aliased. So the fact that  `Order.asc` requires a string is not a problem: it's just what aliases are here for.

If you only had to project on IDs, the code would be like this (excluding the various other calls like `scrollMode` etc.):

```
return criteria.setProjection( Projections.alias( Projections.id(), "aliased_id" ) )
   .addOrder( Order.asc( "aliased_id" ) );
```

So I suppose the issue was that you don't want the result of the query to be the ID, but the whole entity. So you can't simply use `criteria.setProjection( Projections.alias( Projections.id(), "aliased_id" ) )` . And setting up a projection on both the ID and the whole entity indeed seems a bit hard. Judging from the implementation code, I'm not even sure it's possible...

So, plan B: retrieve the ID property name through Hibernate ORM's metadata: `sessionFactory.getClassMetadata( entityType ).getIdentifierPropertyName()`. And put *that* in `idName`.
Should have been plan A, really: it's much simpler. Sorry about that. I get carried away, sometimes :)

By the way, don't forget that `MassIndexerUtil.getIdName` is also used elsewhere. In `org.hibernate.search.jsr352.internal.steps.lucene.PartitionMapper.buildScrollableResults(StatelessSession, Session, Class<?>, Set<Criterion>)` in particular. The issue should be fixed there, too. You could have used aliases there, since you're only projecting on the ID, but sticking to the same solution seems a better idea.

### 3. Side note: `Object.toString()`
To be clear, the issue with using `toString` isn't that it does not suit my taste: it is that the method does not return the string you need. It's like picking a random getter in the class just because it returns a string. From the javadoc of `java.lang.Object#toString`:

> Returns a string representation of the object. In general, the
> `toString` method returns a string that
> "textually represents" this object. The result should
> be a concise but informative representation that is easy for a
> person to read.

... so the intended purpose of this method is to format the object for *humans*, not to serialize it for other parts of the application. In some case it may do both, but `IdentifierProjection.toString` clearly does not.